metrics/src/tests.rs

```diff
@@ -0,0 +1,458 @@
+use std::sync::Arc;
+
+use snafu::{OptionExt, Snafu};
+
+use observability_deps::prometheus::proto::{
+    Counter as PromCounter, Histogram as PromHistogram, MetricFamily,
+};
+
+use crate::MetricRegistry;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("no metric family with name: {}\n{}", name, metrics))]
+    MetricFamilyNotFoundError { name: String, metrics: String },
+
+    #[snafu(display("labels {:?} do not match metric: {}\n{}", labels, name, metrics))]
+    NoMatchingLabelsError {
+        labels: Vec<(String, String)>,
+        name: String,
+        metrics: String,
+    },
+
+    #[snafu(display("metric '{}' failed assertion: '{}'\n{}", name, msg, metrics))]
+    FailedMetricAssertionError {
+        name: String,
+        msg: String,
+        metrics: String,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// A metric registry with handy helpers for asserting conditions on metrics.
+///
+/// You can either initialise a default `TestMetricRegistry` and use the
+/// `registry()` method to inject a `MetricsRegistry` wherever you need it.
+/// Alternatively you can provide an existing `MetricRegistry` via `new`.
+///
+///
+/// The main purpose of the `TestRegistry` is to provide a simple API to assert
+/// that metrics exist and have certain values.
+///
+/// Please see the test cases at the top of this crate for example usage.
+#[derive(Debug, Default)]
+pub struct TestMetricRegistry {
+    registry: Arc<MetricRegistry>,
+}
+
+impl TestMetricRegistry {
+    pub fn new(registry: Arc<MetricRegistry>) -> Self {
+        Self { registry }
+    }
+
+    pub fn registry(&self) -> &MetricRegistry {
+        &self.registry
+    }
+
+    /// Returns an assertion builder for the specified metric name or an error
+    /// if one doesn't exist.
+    ///
+    /// Note: Prometheus histograms comprise multiple metrics sharing the same
+    /// family. Use the family name, e.g., `http_request_duration_seconds` to
+    /// get access to the individual metrics via the `histogram` method on the
+    /// returned `AssertionBuilder`.
+    pub fn try_has_metric_family<'a>(&'a self, name: &str) -> Result<AssertionBuilder<'a>, Error> {
+        let metric_families = self.registry.exporter.registry().gather();
+        let family = metric_families
+            .into_iter()
+            .find(|fam| fam.get_name() == name)
+            .context(MetricFamilyNotFoundError {
+                name: name.to_owned(),
```

Comment:
Function calls eagerly evaluate their arguments, which means that this `to_owned` call will happen unconditionally. However, `context` takes its argument and calls `into` on each of the fields, which allows certain things to behave lazier.

TL;DR:

```suggestion
                name,
```